### PR TITLE
Support dynamic execution for broadcast joins

### PR DESCRIPTION
**Pull request by rschlussel**

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 1]
```diff
@@ -2701,9 +2725,12 @@ public void testGroupedExecution()
             assertQuery(colocatedOneGroupAtATimeDynamic, groupOnJoinResult, expectedGroupOnJoinResult, assertRemoteExchangesCount(2));
 
             assertQuery(broadcastOneGroupAtATime, groupOnJoinResult, expectedGroupOnJoinResult, assertRemoteExchangesCount(2));
+            assertQuery(broadcastOneGroupAtATime, groupOnUngroupedJoinResult, expectedGroupOnUngroupedJoinResult, assertRemoteExchangesCount(2));
+            assertQuery(broadcastOneGroupAtATimeDynamic, groupOnUngroupedJoinResult, expectedGroupOnUngroupedJoinResult, assertRemoteExchangesCount(2));
 
             // cannot be executed in a grouped manner but should still produce correct result
             assertQuery(colocatedOneGroupAtATime, joinUngroupedWithGrouped, expectedJoinUngroupedWithGrouped, assertRemoteExchangesCount(2));
+            assertQuery(colocatedOneGroupAtATime, groupOnUngroupedJoinResult, expectedGroupOnUngroupedJoinResult, assertRemoteExchangesCount(4));
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Why 4 remote exchange ? -- 1 for probe, 1 for build, and 1 for aggregation?

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 2]
```diff
@@ -351,7 +352,7 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
 
                 BucketNodeMap bucketNodeMap;
                 List<Node> stageNodeList;
-                if (plan.getSubStages().isEmpty()) {
+                if (plan.getSubStages().isEmpty() || plan.getFragment().getRemoteSourceNodes().stream().allMatch(node -> node.getExchangeType() == REPLICATE)) {
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: What about 

```java
                if (plan.getFragment().getRemoteSourceNodes().stream().noneMatch(node -> node.getExchangeType() != REPLICATE))
``
?

ditto for the code in `PlanFragmenter.java`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 1 above)
Comment (rschlussel) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: 1 for output

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java (hunk 2 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: thanks! didn't realize that an empty stream would return true, but it makes sense :).

**Comment by wenleix** [comment]
Speaking for the commit title, should it be `Support dynamic lifespan schedule for broadcast joins`?

**Comment by arhimondr** [comment]
Sorry, just got to that. Looks good to me.

**Pull request merged**